internal/topology/variables/clusterclass_variable_validation.go

```diff
@@ -215,9 +225,128 @@ func validateSchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path)
 		allErrs = append(allErrs, validateSchema(&p, fldPath.Child("properties").Key(propertyName))...)
 	}
 
+	// If any schema related validation errors have been found at this level or deeper, skip CEL expression validation.
+	// Invalid OpenAPISchemas are not always possible to convert into valid CEL DeclTypes, and can lead to CEL
+	// validation error messages that are not actionable (will go away once the schema errors are resolved) and that
+	// are difficult for CEL expression authors to understand.
+	if len(allErrs) > 0 {
+		return allErrs
+	}
+
+	celContext := apiextensionsvalidation.RootCELContext(schema)
+	allErrs = append(allErrs, validateCELExpressions(schema, fldPath, celContext)...)
+
+	return allErrs
+}
+
+func validateCELExpressions(schema *apiextensions.JSONSchemaProps, fldPath *field.Path, celContext *apiextensionsvalidation.CELSchemaContext) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if schema.AdditionalProperties != nil {
+		allErrs = append(allErrs, validateCELExpressions(schema.AdditionalProperties.Schema, fldPath.Child("additionalProperties"), celContext.ChildAdditionalPropertiesContext(schema.AdditionalProperties.Schema))...)
+	}
+
+	if len(schema.Properties) != 0 {
+		for property, jsonSchema := range schema.Properties {
+			propertySchema := jsonSchema
+			allErrs = append(allErrs, validateCELExpressions(&propertySchema, fldPath.Child("properties").Key(property), celContext.ChildPropertyContext(&propertySchema, property))...)
+		}
+	}
+
 	if schema.Items != nil {
-		allErrs = append(allErrs, validateSchema(schema.Items.Schema, fldPath.Child("items"))...)
+		allErrs = append(allErrs, validateCELExpressions(schema.Items.Schema, fldPath.Child("items"), celContext.ChildItemsContext(schema.Items.Schema))...)
+		if len(schema.Items.JSONSchemas) != 0 {
+			for i, jsonSchema := range schema.Items.JSONSchemas {
+				itemsSchema := jsonSchema
+				allErrs = append(allErrs, validateCELExpressions(&itemsSchema, fldPath.Child("items").Index(i), celContext.ChildItemsContext(&itemsSchema))...)
+			}
+		}
```

Comment:
```suggestion
```
I would drop it because we never set it

Reply:
Is this worth keeping in case we do set it in future? Would be tricky to find and I don't think this hurts keeping it here even if it is unused.

Reply:
I would prefer dropping it. We have plenty of other cases where we dropped / didn't add a lot of code. There are a few more structs like this that we only ever use one way

Reply:
Dropped at your suggestion.